DXRVanilla/DeusEx/Classes/Player.uc

```diff
@@ -861,9 +863,88 @@ function bool IsFrobbable(actor A)
 
 function HighlightCenterObject()
 {
-    local Vector loc;
+    HighlightCenterObjectMain();
+    HighlightCenterObjectLaser();
+}
+
+function HighlightCenterObjectMain()
+{
+    local Actor target, smallestTarget;
+    local Vector HitLoc, HitNormal, StartTrace, EndTrace;
+    local DeusExRootWindow root;
+    local float minSize;
+    local bool bFirstTarget;
+
+    if (IsInState('Dying'))
+        return;
 
-    Super.HighlightCenterObject();
+    root = DeusExRootWindow(rootWindow);
+
+    // only do the trace every hundredth of a second, this is 10 times faster than vanilla
+    if (FrobTime >= 0.01)
+    {
+        // figure out how far ahead we should trace
+        StartTrace = Location;
+        EndTrace = Location + (Vector(ViewRotation) * MaxFrobDistance);
+
+        // adjust for the eye height
+        StartTrace.Z += BaseEyeHeight;
+        EndTrace.Z += BaseEyeHeight;
+
+        smallestTarget = None;
+        minSize = 99999;
+        bFirstTarget = True;
+
+        // find the object that we are looking at
+        // make sure we don't select the object that we're carrying
+        // use the last traced object as the target...this will handle
+        // smaller items under larger items for example
+        // ScriptedPawns always have precedence, though
+        foreach TraceActors(class'Actor', target, HitLoc, HitNormal, EndTrace, StartTrace)
+        {
+            if (IsFrobbable(target) && (target != CarriedDecoration))
+            {
+                if (target.IsA('ScriptedPawn'))
+                {
+                    smallestTarget = target;
+                    break;
+                }
+                else if (target.IsA('Mover') && bFirstTarget)
+                {
+                    smallestTarget = target;
+                    break;
+                }
+                else if (target.CollisionRadius < minSize)
+                {
+                    minSize = target.CollisionRadius;
+                    smallestTarget = target;
+                    bFirstTarget = False;
+                }
+            }
+        }
+
+        // if we already have a frob target, and the player looks away such as that no item is being
+        // traced, we still wait for the full 100ms vanilla duration before clearing the frob
+        // target.
+        //
+        // note that this means we don't wait for the full 100ms vanilla duration if the player is
+        // rapidly changing frob target.
+        if ((FrobTime < 0.1) && (FrobTargetFromHighlight != None) && (smallestTarget == None))
+        {
+            return;
+        }
+
+        FrobTarget = smallestTarget;
+        FrobTargetFromHighlight = smallestTarget;
```

Comment:
I'm not sure `FrobTargetFromHighlight` is useful, it seems the same as FrobTarget, if other code sets FrobTarget to None then you're not highlighting anything and maybe we should respect that, but I haven't tested it

Reply:
When I have time I'll remove it and do some more thorough testing, which may be as soon as tomorrow.